### PR TITLE
refactor(actions): rename workflow files

### LINKED ISSUE #3924: actions: rename workflows to correctly identify what's happening in it
## Motivation

Most of the steps being executed in `ci.yml` are basically unit tests and the ones being done in `test.yml` are integration tests
This sometimes creates confusion for most people when talking about CI, as the actual Continuos Integration is being done in `test.yml` and not in `ci.yml`

Standardize other names accordingly.

### Designs
- Change `test.yml` to `integration-tests.yml`
- Change `ci.yaml` to `unit-tests.yml`
- Change `test-full-sync` to `full-sync-tests.yml`
- Change `cd.yaml` to `continous-delivery.yml`

## Related Work

Further work will be done for a `continous-deployment.yml` workflow

### PR DESCRIPTION
**Pull request by gustavovalverde**
## Motivation

Most of the steps being executed in `ci.yml` are basically unit tests and the ones being done in `test.yml` are integration tests
This sometimes creates confusion for most people when talking about CI, as the actual Continuos Integration is being done in `test.yml` and not in `ci.yml`

Fixes: #3924
Depends-On: #3940 

Standardize other names accordingly.

## Solution

- Change `cd.yaml` to `continous-delivery.yml`
- Change `test.yml` to `continous-integration.yml`
- Change `ci.yaml` to `os-build-tests.yml`
- Change `test-full-sync` to `full-sync-tests.yml`

## Review

@dconnolly or @teor2345 can review this

### CODE REVIEW

**Review by teor2345 — DISMISSED** [review]
I think this PR might change CI behaviour, because some of these workflow files contain references to their own names.

Does GitHub have a way of saying "run this workflow if the workflow itself changes"?

If not, can you do a search and replace for all these name changes? (I recommend `fastmod`, it has a great interactive search and replace mode.)

Optional: Could we name all the test workflows starting with "test", so they are listed together in the workflow directory?

(I really don't mind either way, but it makes them a bit easier to find.)

**Comment by gustavovalverde** [comment]
@teor2345 @dconnolly I think we should remove `test-fake-activation-heights` from here https://github.com/ZcashFoundation/zebra/blob/9e0bbb5a6a41ee9549a525a45f21acf3ef3ff9c6/.github/workflows/os-build-tests.yml#L128 as we already have it here https://github.com/ZcashFoundation/zebra/blob/9e0bbb5a6a41ee9549a525a45f21acf3ef3ff9c6/.github/workflows/continous-integration.yml#L161

And this not OS-dependent, as far as I know.

**Comment by teor2345** [comment]
> @teor2345 @dconnolly I think we should remove `test-fake-activation-heights` from here
> 
> And this not OS-dependent, as far as I know.

It only needs to run on one OS.
Can you make that change in a separate PR?

**Comment by gustavovalverde** [comment]
> Optional: Could we name all the test workflows starting with "test", so they are listed together in the workflow directory?

@teor2345 just added this recommendation

**Review by teor2345 — COMMENTED** [review]
We have a very tight NU5 release deadline - we need to tag by Tuesday 19 April.

Can we hold this risky PR until after we've tagged that release?

Do we need to merge this PR manually, so we can update the GitHub branch protection rules right after it?

File: .github/workflows/pr-garbage-collection.yml [hunk 1]
```diff
@@ -1,4 +1,4 @@
-name: Clean
+name: Merge cleanup
```
Comment (teor2345) [inline .github/workflows/pr-garbage-collection.yml]: I'm not sure about the names here, can we use something more specific?
And can we make the workflow and file names similar? 

```suggestion
name: Clean up Google Cloud instances
```

File: .github/workflows/continous-integration.yml [hunk 2]
```diff
@@ -1,4 +1,4 @@
-name: Test
+name: CI
```
Comment (teor2345) [inline .github/workflows/continous-integration.yml]: Can we make the Docker and OS names more consistent?
And start both files with the same word?

```suggestion
name: Test on Docker
```

File: .github/workflows/test-os-build.patch.yml [hunk 3]
```diff
@@ -1,4 +1,4 @@
-name: CI
+name: Test OS builds
```
Comment (teor2345) [inline .github/workflows/test-os-build.patch.yml]:
```suggestion
name: Test on OSes
```

**Review by teor2345 — COMMENTED** [review]
I'd like the names to be more consistent, because that's what I look for when I search for the right workflow.

**Comment by gustavovalverde** [comment]
I'm not completely sure on renaming `CI` to `Test on Docker`. My opinions on that:

- I started changing the names to have a broad (and kind of standard) understanding on what's happening on the pipelines. And as a general rule of thumb I decided to use the DevOps segmentation: Continuos Integration, Continuos Delivery and Continuos Deployment (still missing).

- If we remove `test-fake-activation-heights` from the `OS tests` workflow, we will only be testing if the build is successful on different OSes (I've just realized that we should also remove `sync_large_checkpoints_` and `--all` tests to make this change consistent, as those are already available at `continous-integration.yml`)

- If we're going to make tests on the OSes and not just builds for different OSes, then we have two roads for standardization: Adding the missing tests from the `continous-integration.yml` to `test-os-build.yml` or completely remove the tests from `test-os-build.yml` and simply do build related steps.

If we agree on doing `integration tests` in the different OSes (not just `build` steps) then we might be better merging everything in the `continous-integration.yml` or having `continous-integration-docker.yml` and `continous-integration-os.yml` (or similar)


**Related**

This decision is also related to the level of support of the different OSes. Which we recently agreed it would be:
- **TIER 1/deployment/run support** (“guaranteed to work”) for Linux x64, mainly with Debian Bullseye.
- **TIER 2/build support** (“guaranteed to build”) for Linux and Darwin.
- **TIER 3/codebase support** (“may or may not work”) for Windows.

To help on this decision, the pipelines should be aligned to these support levels.

**Comment by teor2345** [comment]
> I'm not completely sure on renaming `CI` to `Test on Docker`. My opinions on that:
> * I started changing the names to have a broad (and kind of standard) understanding on what's happening on the pipelines. And as a general rule of thumb I decided to use the DevOps segmentation: Continuos Integration, Continuos Delivery and Continuos Deployment (still missing).

This naming scheme sounds good to me!

> * If we remove `test-fake-activation-heights` from the `OS tests` workflow, we will only be testing if the build is successful on different OSes (I've just realized that we should also remove `sync_large_checkpoints_` and `--all` tests to make this change consistent, as those are already available at `continous-integration.yml`)

Disabling tests is out of scope for a renaming PR, but I understand that the discussion might be helpful to choose names.

I'm not sure if this test matrix is what we agreed on. If we want to drop all the non-Docker tests, that seems like a decision the team needs to make.

I am also open to disabling redundant tests that should not vary by OS, but that's a functionality change. If we do it in a separate PR or ticket, we can prioritise it better.

> * If we're going to make tests on the OSes and not just builds for different OSes, then we have two roads for standardization: Adding the missing tests from the `continous-integration.yml` to `test-os-build.yml` or completely remove the tests from `test-os-build.yml` and simply do build related steps.

I think we're missing an alternative here:
* do some tests on tier 2 platforms, and all the tests on tier 1 platforms

This is what we currently do, and I'd like to keep this behaviour in this PR.

> If we agree on doing `integration tests` in the different OSes (not just `build` steps) then we might be better merging everything in the `continous-integration.yml` or having `continous-integration-docker.yml` and `continous-integration-os.yml` (or similar)

I would prefer `continous-integration-docker.yml` and `continous-integration-os.yml`, so we can:
* see which tier is failing
* temporarily or permanently disable a lower priority tier using the GitHub actions settings

I'd also like to move the `cargo deny` checks into the lints, because that's what they are. But that can happen in a separate PR.

**Comment by teor2345** [comment]
I split this into a separate comment, because the previous one was getting a bit long.

> **Related**
> This decision is also related to the level of support of the different OSes. Which we recently agreed it would be:
> * **TIER 1/deployment/run support** (“guaranteed to work”) for Linux x64, mainly with Debian Bullseye.
> * **TIER 2/build support** (“guaranteed to build”) for Linux and Darwin.
> * **TIER 3/codebase support** (“may or may not work”) for Windows.
> To help on this decision, the pipelines should be aligned to these support levels.

I agree in principle, but I'm not sure if these are the exact details we agreed on.

Here's what I was thinking after our recent discussion:
* **TIER 1/deployment/run support** (“guaranteed to work”) for Linux _x86-_ 64, ~mainly with~ Debian Bullseye *on Docker*.
* **TIER 2/build support** (“guaranteed to build”) for *mainstream* Linux and *recent macOS on x86-64*.
* **TIER 3/codebase support** (“may or may not work”) for Windows, *and other OSes and processors, including non-standard Linux and Darwin distributions*.

Also, the tiers are *minimum* support levels, so we are free to do extra checks if we want. But we decided not to put much effort into them. So if they stop working, we should just turn them off.

Did we document those support levels anywhere?
It would help to have a single document the team can agree on.
Is that something you could do?

**Comment by teor2345** [comment]
@gustavovalverde I realise support levels are a big conversation, so let me know if you want to catch up 1:1 or with the team?

And if we have to tweak the workflow names again later, that's not a big deal.

File: .github/workflows/lint.yml [hunk 4]
```diff
@@ -1,4 +1,4 @@
-name: Lint Rust files
+name: Linter
```
Comment (dconnolly) [inline .github/workflows/lint.yml]: 👍

**Comment by teor2345** [comment]
Not required to merge PR #4267, which is our top priority right now.

**Comment by gustavovalverde** [comment]
@dconnolly I think we're ready with this one, I've renamed some files to have the similar ones grouped together

**Comment by gustavovalverde** [comment]
Visual reference:
image

**Comment by dconnolly** [comment]
This looks close, I wonder if we can distinguish `-docker.yml` and `-os.yml` a little more about what they do / their goals, an idea:

`continuous integration requiring cloud infra`
`continuous integration` (implicitly without gcloud infra, i guess i'm agnostic to explicitly saying 'for os testing' or something)

**Comment by gustavovalverde** [comment]
I think that something is _clear_ here:
- `continous-integration-docker.yml` runs all integration tests, those being locally run or in the infra. IMO we can keep just as `continous-integration.yml`. Adding `-docker.yml` might be over explicit, as (so far) all of our deployment, test or not, are dockerized.

To be honest, `-os.yml` is difficult one to decide because:
- It's `building` for different OSes (the main goal as I see it)
- But as I discussed with Teor here https://github.com/ZcashFoundation/zebra/pull/3941#issuecomment-1102186264, it also runs two (2) integration tests.

We have 3 available decisions to make (out of the scope of this PR):
- Leave it as `continous-integration-os.yml` and `continous-integration-docker.yml` in the meanwhile
- Create another PR and decide if:
  - We add all missing tests from `continous-integration-docker.yml` into `continous-integration-os.yml`
  - We remove all the tests from `continous-integration-os.yml` (`test-fake-activation-heights`, `sync_large_checkpoints_` and `--all`) and leave it as a simple `test-os-builds.yml`
  

I'd choose to keep as it is right now, until we decide or discuss if the actual tests being run in `continous-integration-os.yml` should be removed, kept or expanded.

**Comment by dconnolly** [comment]
> I think that something is _clear_ here:
> 
> * `continous-integration-docker.yml` runs all integration tests, those being locally run or in the infra. IMO we can keep just as `continous-integration.yml`. Adding `-docker.yml` might be over explicit, as (so far) all of our deployment, test or not, are dockerized.
> 
> To be honest, `-os.yml` is difficult one to decide because:
> 
> * It's `building` for different OSes (the main goal as I see it)
> * But as I discussed with Teor here [refactor(actions): rename workflow files #3941 (comment)](https://github.com/ZcashFoundation/zebra/pull/3941#issuecomment-1102186264), it also runs two (2) integration tests.
> 
> We have 3 available decisions to make (out of the scope of this PR):
> 
> * Leave it as `continous-integration-os.yml` and `continous-integration-docker.yml` in the meanwhile
> * Create another PR and decide if:
>   
>   * We add all missing tests from `continous-integration-docker.yml` into `continous-integration-os.yml`
>   * We remove all the tests from `continous-integration-os.yml` (`test-fake-activation-heights`, `sync_large_checkpoints_` and `--all`) and leave it as a simple `test-os-builds.yml`
> 
> I'd choose to keep as it is right now, until we decide or discuss if the actual tests being run in `continous-integration-os.yml` should be removed, kept or expanded.

Let's go with the wording in here for now 👍

**Review by dconnolly — APPROVED** [review]

**Pull request merged**